src/libcore/iter.rs

```diff
@@ -2421,9 +2421,24 @@ macro_rules! step_impl {
             }
             #[inline]
             #[allow(trivial_numeric_casts)]
+            #[allow(unused_comparisons)]
             fn steps_between(start: &$t, end: &$t, by: &$t) -> Option<usize> {
-                if *start <= *end {
-                    Some(((*end - *start) / *by) as usize)
+                if *by == 0 { return None; }
+                let mut start = *start;
+                let mut end = *end;
+                let mut by = *by;
+                if by < 0 {
+                    start *= -1;
+                    end *= -1;
+                    by *= -1;
+                }
```

Comment:
Was this block just necessary for the `%` check below? It looks like otherwise the signs just end up handling themselves.